video/out/vo_drm.c

```diff
@@ -0,0 +1,509 @@
+/*
+ * video output driver for libdrm
+ *
+ * by rr- <rr-@sakuya.pl>
+ *
+ * This file is part of mpv.
+ *
+ * mpv is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * mpv is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <stdbool.h>
+#include <libswscale/swscale.h>
+#include <sys/mman.h>
+#include <xf86drm.h>
+#include <xf86drmMode.h>
+
+#include "common/msg.h"
+#include "sub/osd.h"
+#include "video/fmt-conversion.h"
+#include "video/mp_image.h"
+#include "video/sws_utils.h"
+#include "vo.h"
+
+#define BUF_COUNT 2
+
+struct modeset_buf {
+    uint32_t width;
+    uint32_t height;
+    uint32_t stride;
+    uint32_t size;
+    uint32_t handle;
+    uint8_t *map;
+    uint32_t fb;
+};
+
+struct modeset_dev {
+    struct modeset_buf bufs[BUF_COUNT];
+    drmModeModeInfo mode;
+    uint32_t conn;
+    uint32_t crtc;
+    int front_buf;
+};
+
+struct priv {
+    int fd;
+    struct modeset_dev *dev;
+    drmModeCrtc *old_crtc;
+
+    char *device_path;
+    int connector_id;
+
+    int32_t device_w;
+    int32_t device_h;
+    int32_t x, y;
+    struct mp_image *curframe;
+    struct mp_rect src;
+    struct mp_rect dst;
+    struct mp_osd_res osd;
+    struct mp_sws_context *sws;
+};
+
+static int modeset_open(struct vo *vo, int *out, const char *node)
+{
+    int fd;
+    uint64_t has_dumb;
+    fd = open(node, O_RDWR | O_CLOEXEC);
+    if (fd < 0) {
+        MP_ERR(vo, "Cannot open \"%s\": %s.\n", node, mp_strerror(errno));
+        return -errno;
+    }
+    if (drmGetCap(fd, DRM_CAP_DUMB_BUFFER, &has_dumb) < 0) {
+        MP_ERR(vo, "Device \"%s\" does not support dumb buffers.\n", node);
+        return -EOPNOTSUPP;
+    }
+    *out = fd;
+    return 0;
+}
+
+static int modeset_create_fb(struct vo *vo, int fd, struct modeset_buf *buf)
+{
+    struct drm_mode_create_dumb creq;
+    struct drm_mode_destroy_dumb dreq;
+    struct drm_mode_map_dumb mreq;
+    int ret;
+
+    // create dumb buffer
+    memset(&creq, 0, sizeof(creq));
```

Comment:
Instead of this, you can just write ``struct drm_mode_create_dumb creq = {0};``.